[core] Refactor how we store data in redis.

rynewang opened the pull request:
Ray GCS Fault Tolerance replies on an external Redis. We write all data for a Ray cluster into a single Redis HASH. Within the HASH, we use a Redis field prefix to represent a Ray table. This is OK for single field CRUD, but for "get all entries for a Ray table" we would have to do a full table HSCAN and filter out other Ray tables by key prefix. This slows things down and the call to get_all_node_info and get_all_actor_info are in fact fairly common.

Updates the Redis usage by creating multiple Redis HASHes, 1 for each Ray table. The single field CRUD are largely the same; but "get all" operations now become a simple HSCAN on the target HASH only. This saves lots of compute.

Closes #46706.

rynewang commented:
ping @rkooo567

rkooo567 commented:
can you link  the original issue?

rynewang commented:
There's no issue; it's from a community PR #46706

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -83,55 +64,38 @@ std::vector<std::vector<std::string>> GenCommandsBatched(
-            (*key_value_map)[GetKeyFromRedisKey(
-                external_storage_namespace_, mget_keys[index + 2], table_name)] =
-                *(value[index]);
+    std::vector<std::string> partition_keys(command.begin() + 2, command.end());
```
what is + 2 for? can you comment to c ode?

rynewang replied:
added to comment

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -83,55 +64,38 @@ std::vector<std::vector<std::string>> GenCommandsBatched(
-                *(value[index]);
+    std::vector<std::string> partition_keys(command.begin() + 2, command.end());
+    auto mget_callback =
+        [finished_count, total_count, partition_keys, callback, key_value_map](
```
nit: move callback? (callback=std::move(callback))

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -500,6 +449,62 @@ Status RedisStoreClient::AsyncExists(const std::string &table_name,
+
+  auto context = cli->GetPrimaryContext();
+  auto cmd = std::vector<std::string>{"KEYS", GenRedisKeyPrefixMatchPattern(key_prefix)};
+  auto reply = context->RunArgvSync(cmd);
```
do we expect to copy the output?

rynewang replied:
good spot! fixed by adding ref.

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -500,6 +449,62 @@ Status RedisStoreClient::AsyncExists(const std::string &table_name,
+  auto context = cli->GetPrimaryContext();
+  auto cmd = std::vector<std::string>{"KEYS", GenRedisKeyPrefixMatchPattern(key_prefix)};
+  auto reply = context->RunArgvSync(cmd);
+  const auto keys = reply->ReadAsStringArray();
```
```suggestion
  const auto &keys = reply->ReadAsStringArray();
```

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.h:
```diff
@@ -166,6 +185,13 @@ class RedisStoreClient : public StoreClient {
 };
 
+// Helper function used by Python to delete all redis HASHes with a given prefix.
+bool RedisDelKeyPrefixSync(const std::string &host,
```
can you add a docstring for the return value

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -83,55 +64,38 @@ std::vector<std::vector<std::string>> GenCommandsBatched(
-    };
-    SendRedisCmd(
-        std::move(partition_keys), std::move(mget_keys), std::move(mget_callback));
+          ++(*finished_count);
```
maybe not a scope of this PR, but how do we handle timeout in this case?

rkooo567 replied:
(in that case it hangs forever?)

rynewang replied:
I don't think we have a timeout config now. We can use `redisAsyncSetTimeout` to do it but nobody cares. BTW GCS assumes Redis is always available so we don't really can do anything on a redis timeout.

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -381,28 +325,27 @@ Status RedisStoreClient::DeleteByKeys(const std::vector<std::string> &keys,
-      external_storage_namespace_(external_storage_namespace),
+      match_pattern_(match_pattern),
       redis_client_(std::move(redis_client)) {
+  RAY_CHECK(!match_pattern.empty());
```
add an error msg?

rkooo567 replied:
"match_pattern is not given. It should look like x"

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -415,26 +358,23 @@ void RedisStoreClient::RedisScanner::Scan(const std::string &match_pattern,
-      [this, match_pattern, callback](const std::shared_ptr<CallbackReply> &reply) {
-        OnScanCallback(match_pattern, reply, callback);
-      };
+  auto scan_callback = [this, callback](const std::shared_ptr<CallbackReply> &reply) {
```
can we move callback?

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -415,26 +358,23 @@ void RedisStoreClient::RedisScanner::Scan(const std::string &match_pattern,
-                                   match_pattern,
-                                   "COUNT",
-                                   std::to_string(batch_count)};
+  std::vector<std::string> args = {"HSCAN", table_name_, std::to_string(cursor_.value())};
```
why does it not use GenRedisKey?

rynewang replied:
Ctor of RedisScanner accepts the table name, and all its callers uses GenRedisKey so we are fine.

rynewang replied:
Do you want a more typed set up, e.g. a struct represending the {external_storage_namespace, table_name} ?

rkooo567 replied:
> Do you want a more typed set up, e.g. a struct represending the {external_storage_namespace, table_name} ?

Yeah this would be great. I am also confused why this doesn't include `external_storage_namespace` (that's why I was asking why we don't use GenRedisKey)?

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -448,50 +388,59 @@ void RedisStoreClient::RedisScanner::OnScanCallback(
-      auto key = GetKeyFromRedisKey(
-          external_storage_namespace_, std::move(scan_result[i]), table_name_);
-      results_.emplace(std::move(key), std::move(scan_result[i + 1]));
+      results_.emplace(std::move(scan_result[i]), std::move(scan_result[i + 1]));
```
can you tell me a little bit about why this is changed?

rynewang replied:
Previously:
- req: `HSCAN hash_with_cluster_id MATCH Jobs*`
- resp: `Jobs@job1 job1_value`
- GetKeyFromRedisKey extracts: `job1`
- Saves to results_: `job1: job1_value`

Now:
- req: `HSCAN hash_with_cluster_id_for_Jobs`
- resp: `job1 job1_value`
- no need to call GetKeyFromRedisKey, key is directly job_id (job1)
- Saves to results_: `job1: job1_value`

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -448,50 +388,59 @@ void RedisStoreClient::RedisScanner::OnScanCallback(
+  std::vector<std::string> args = {"INCRBY", key, "1"};
+
+  auto cxt = redis_client_->GetPrimaryContext();
+  auto reply = cxt->RunArgvSync(args);
```
```suggestion
  const auto &reply = cxt->RunArgvSync(args);
```
?

rynewang replied:
RunArgvSync returns unique_ptr so it's good to keep it by value.

rkooo567 commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -500,6 +449,62 @@ Status RedisStoreClient::AsyncExists(const std::string &table_name,
+  });
+
+  Cleanup _([&]() {
+    io_service.stop();
```
this assumes io_service is GC'ed after Cleanup object. Is it true? maybe we should make io_service a shared pointer?

rkooo567 replied:
same for thread

rynewang replied:
yes this is safe. dtor call order is in reverse of ctor call order. so

io_service ctor
thread ctor
Cleanup ctor

// work

Cleanup dtor (stops io_service, joins thread)
thread dtor
io_service dtor

so we are safe.

rkooo567 reviewed (commented):
Looks good in general. many minor comments.

Besides, how do we test them now?

rynewang commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.md:
```diff
@@ -141,7 +141,7 @@ KEYS *
 
 # Step 6.5: Check the value of the key.
-HGETALL 864b004c-6305-42e3-ac46-adfa8eb6f752
+SCAN 0 MATCH 864b004c-6305-42e3-ac46-adfa8eb6f752*
```
Do we expect users to directly operate ray redis?

kevin85421 replied:
I think this change is fine. My only concern is about cleaning up Redis. @rynewang will provide me with a Docker image, and I will manually test it.

rynewang replied:
Kuberay cleans redis by calling del_key_prefix_from_storage which I reimplemented so should be OK. I'm giving @kevin85421 a docker image to test

kevin85421 replied:
It's not easy to get a Docker image. Let's merge it and then I will use the nightly image to test.

rkooo567 replied:
Also cc @edoakes we are using the function to clean up redis or the raw command?

jjyao replied:
Yea, I checked with @edoakes before, we also use `cleanup_redis_storage`

rynewang commented:
@rkooo567 PTAL again

rkooo567 commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.md:
```diff
@@ -141,7 +141,7 @@ KEYS *
 
 # Step 6.5: Check the value of the key.
-HGETALL 864b004c-6305-42e3-ac46-adfa8eb6f752
+SCAN 0 MATCH 864b004c-6305-42e3-ac46-adfa8eb6f752*
```
can you mention which release this command is changed from here?

rynewang replied:
wdym "which release"?

rkooo567 replied:
like it is changed from ray 2.35 or sth like that.

rynewang replied:
done

rkooo567 reviewed (approved):
Looks good. Last couple comments (please make sure to resolve)!

Regarding Redis, can you sync with @kevin85421 and make sure we recommend to use `cleanup_redis_storage `, not the raw Redis command? If we have any recommendation using raw Redis command, I think we should have a loud release note.

rynewang commented:
Rewrote most of the file to use typed RedisCommand, RedisKey and RedisMatchPattern. PTAL! @rkooo567

rkooo567 reviewed (approved):
looks very good!

rynewang commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.md:
```diff
@@ -141,7 +141,7 @@ KEYS *
 
 # Step 6.5: Check the value of the key.
-HGETALL 864b004c-6305-42e3-ac46-adfa8eb6f752
+SCAN 0 MATCH 864b004c-6305-42e3-ac46-adfa8eb6f752*
```
you r right, will update

rynewang commented on src/ray/gcs/store_client/redis_store_client.h:
```diff
@@ -30,6 +30,60 @@ namespace ray {
+//
+// Schema:
+// - Each table is a Redis HASH. The HASH "key" is
+//    "RAY" + `external_storage_namespace` + "@" + `table_name`.
```
No, now it's just external_storage_namespace_

rynewang commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -148,10 +117,19 @@ Status RedisStoreClient::AsyncPut(const std::string &table_name,
+          callback(added_num != 0);
+        };
+  }
+  SendRedisCmd({key}, std::move(command), std::move(write_callback));
```
will change

rynewang replied:
done

jjyao commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -277,8 +249,8 @@ std::vector<std::function<void()>> RedisStoreClient::TakeRequestsFromSendingQueu
 
-void RedisStoreClient::SendRedisCmd(std::vector<std::string> keys,
-                                    std::vector<std::string> args,
+void RedisStoreClient::SendRedisCmd(std::vector<RedisConcurrencyKey> keys,
```
I think we can still accept the original `std::vector<std::string> keys`, concurrency key is just `{command.redis_key, key}`

jjyao replied:
In a follow-up PR: we should rename keys to fields which is the term used by redis

rynewang replied:
done, removed keys even.

rynewang commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.md:
```diff
@@ -136,12 +136,22 @@ export REDIS_POD=$(kubectl get pods --selector=app=redis -o custom-columns=POD:m
 # Step 6.5: Check the value of the key.
-HGETALL 864b004c-6305-42e3-ac46-adfa8eb6f752
+HGETALL RAY864b004c-6305-42e3-ac46-adfa8eb6f752@NODE
+# Before Ray 2.35.0:
```
changed to 2.36

jjyao commented on src/ray/gcs/store_client/redis_store_client.h:
```diff
@@ -130,27 +216,32 @@ class RedisStoreClient : public StoreClient {
+  // Reason for this method: if you call SendRedisCmd(command.args, std::move(command)),
+  // it's UB because C++ don't have arg evaluation order guarantee, hence command.args
+  // may become empty.
+  void SendRedisCmd(RedisCommand command, RedisCallback redis_callback);
```
```suggestion
  void SendRedisCmdArgsAsKeys(RedisCommand command, RedisCallback redis_callback);
```

I'm a bit worried that users will misuse these two functions without reading the comment. This rename should make it a bit clear.

Or should we still just have a single version of SendRedisCmd and force users to always specify keys (they need to manually do what SendRedisCmd does in this PR)?

rynewang replied:
updated

jjyao commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -166,28 +145,25 @@ Status RedisStoreClient::AsyncGet(const std::string &table_name,
-  std::string redis_key = GenRedisKey(external_storage_namespace_, table_name, key);
-  std::vector<std::string> args = {"HGET", external_storage_namespace_, redis_key};
-  SendRedisCmd({redis_key}, std::move(args), std::move(redis_callback));
+  RedisCommand command{.command = "HGET",
+                       .redis_key = RedisKey{external_storage_namespace_, table_name},
+                       .args = {key}};
```
windows will unhappy?

rynewang replied:
good spot, updated

jjyao commented on src/ray/gcs/store_client/redis_store_client.h:
```diff
@@ -30,6 +30,78 @@ namespace ray {
+  std::vector<std::string> args;
+
+  std::vector<std::string> ToRedisArgs() const {
+    std::vector<std::string> redis_args;
```
nit: we know the size beforehand so we can reserve

jjyao commented:
I think the test failure is real (i.e. cannot be fixed by increasing timeout)

rynewang commented:
it IS real. It happens when you have a large BatchDelete, where a bad move on callback function makes the callback never called. fixed and reverted the test relaxings.

jjyao commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.md:
```diff
@@ -136,12 +136,22 @@ export REDIS_POD=$(kubectl get pods --selector=app=redis -o custom-columns=POD:m
 kubectl exec -it $REDIS_POD -- redis-cli -a "5241590000000000"
 
 # Step 6.4: Check the keys in Redis.
+# Note: the schema changed in Ray 2.36.0. Previously we use a single HASH table,
```
This needs to be updated to 2.38.0.

angelinalg reviewed (approved):
stamp

jjyao commented on python/ray/_private/gcs_utils.py:
```diff
@@ -142,6 +142,9 @@ def cleanup_redis_storage(
-    # So we only need to delete the specific key to cleanup the cluster.
-    return del_key_from_storage(host, port, password, use_ssl, storage_namespace)
+    # Right now, GCS stores all data into multiple hashes with keys prefixed by
+    # storage_namespace. So we only need to delete the specific key to cleanup the
+    # cluster.
```
This is no longer accurate? We need to delete multiple keys now.

jjyao commented on python/ray/_private/gcs_utils.py:
```diff
@@ -142,6 +142,9 @@ def cleanup_redis_storage(
+    # Right now, GCS stores all data into multiple hashes with keys prefixed by
+    # storage_namespace. So we only need to delete the specific key to cleanup the
+    # cluster.
+    return del_external_storage_namespace_from_storage(
```
I feel the old name is cleaner: `del_key_prefix_from_storage`. Or `del_keys_with_prefix_from_storage` is even better IMO.

rynewang replied:
I wanted to hide the key schema (that external_storage_namespace is prefix of the keys). But ok I can change that back to prefix.

jjyao replied:
Got it. Then I think we don't have to make the name generic, it can be just `cleanup_redis_storage()`. I'm merging for now and we can rename it later.

JoshKarpel commented:
Hi, just spotted this - wondering if corresponding changes are being made to KubeRay to ensure that it is able to clean up all of these new tables in the GCS cleanup job that runs when a cluster is terminated?

Alternatively, it would be nice if these tables had expirations set (and regularly refreshed) internally so that KubeRay didn't need to clean them up.

See https://github.com/ray-project/kuberay/issues/959

rynewang commented:
Yes, KubeRay calls a Ray function to clean up tables and that function is
updated accordingly. For the table expiration - that’s something separate
and yes it’s a good thing to have, though not in this PR.

On Wed, Sep 18, 2024 at 16:43 Josh Karpel ***@***.***> wrote:

> Hi, just spotted this - wondering if corresponding changes are being made
> to KubeRay to ensure that it is able to clean up all of these new tables in
> the GCS cleanup job that runs when a cluster is terminated?
>
> Alternatively, it would be nice if these tables had expirations set (and
> regularly refreshed) internally so that KubeRay didn't need to clean them
> up.
>
> See ray-project/kuberay#959
> <https://github.com/ray-project/kuberay/issues/959>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/46861#issuecomment-2359633798>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANLX3X4MDUJR2463T43B34DZXIF3JAVCNFSM6AAAAABLVZAYHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNJZGYZTGNZZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JoshKarpel commented:
Awesome, thank you!
